graph-commons/src/test/scala/io/renku/http/client/GitLabClientSpec.scala

```diff
@@ -0,0 +1,177 @@
+package io.renku.http.client
+
+import cats.effect.IO
+import cats.implicits.{catsSyntaxApplicativeErrorId, toShow}
+import com.github.tomakehurst.wiremock.client.MappingBuilder
+import com.github.tomakehurst.wiremock.client.WireMock._
+import io.renku.control.Throttler
+import io.renku.generators.CommonGraphGenerators.{oauthAccessTokens, pages, personalAccessTokens}
+import io.renku.generators.Generators.Implicits._
+import io.renku.generators.Generators._
+import io.renku.graph.model.GitLabUrl
+import io.renku.http.client.RestClientError.{UnauthorizedException, UnexpectedResponseException}
+import io.renku.interpreters.TestLogger
+import io.renku.logging.TestExecutionTimeRecorder
+import io.renku.metrics.GitLabApiCallRecorder
+import io.renku.stubbing.ExternalServiceStubbing
+import io.renku.testtools.IOSpec
+import org.http4s.Method._
+import org.http4s.Status.InternalServerError
+import org.http4s.{Method, Uri}
+import org.scalacheck.Gen
+import org.scalamock.scalatest.MockFactory
+import org.scalatest.matchers.should
+import org.scalatest.wordspec.AnyWordSpec
+
+class GitLabClientSpec
+    extends AnyWordSpec
+    with IOSpec
+    with ExternalServiceStubbing
+    with should.Matchers
+    with MockFactory {
+
+  "send" should {
+
+    "fetch json from a given page" in new TestCase {
+
+      stubFor {
+        callMethod(method, s"api/v4/$path")
+          .willReturn(
+            okJson(result)
+              .withHeader("X-Next-Page", maybeNextPage.map(_.show).getOrElse(""))
+          )
+      }
+
+      client.send(method, path, endpointName)(_ => IO(result)).unsafeRunSync() shouldBe result
+    }
+
+    "fetch commits using the given personal access token" in new TestCase {
+      val personalAccessToken = personalAccessTokens.generateOne
+
+      stubFor {
+        callMethod(method, s"api/v4/$path")
+          .withHeader("PRIVATE-TOKEN", equalTo(personalAccessToken.value))
+          .willReturn(
+            okJson(result)
+              .withHeader("X-Next-Page", maybeNextPage.map(_.show).getOrElse(""))
+          )
+      }
+
+      client.send(method, path, endpointName)(_ => IO(result)).unsafeRunSync() shouldBe result
+    }
+
+    "fetch commits using the given oauth token" in new TestCase {
+      val authAccessToken = oauthAccessTokens.generateOne
+
+      stubFor {
+        callMethod(method, s"api/v4/$path")
+          .withHeader("PRIVATE-TOKEN", equalTo(authAccessToken.value))
+          .willReturn(
+            okJson(result)
+              .withHeader("X-Next-Page", maybeNextPage.map(_.show).getOrElse(""))
+          )
+      }
+
+      client.send(method, path, endpointName)(_ => IO(result)).unsafeRunSync() shouldBe result
+    }
+
+    "Handle empty JSON" in new TestCase {
```

Comment:
I'm afraid this test and the subsequent three are not achieving what they suppose to. The reason for that is the `mappingResponse` function that is passed to the send method does not use/interact with the response. I believe the idea here is to give some mapping function that tries, for instance in this particular test, to decode some value from the response (in other cases fail if the response's status is different than the status in the mapping function, and so on).